POEM_002.md

```diff
@@ -0,0 +1,86 @@
+
+Status:
+
+- [x] Active
+- [ ] Requesting decision
+- [ ] Accepted
+- [ ] Rejected
+- [ ] Integrated
+
+
+Motivation
+==========
+The SNOPT optimizer provides a way to cleanly terminate an optimization by returning a status of -2
+or lower from the model function and gradient evaluation functions. This allows for a clean and
+graceful exit from optimization. To take advantage of this under a wide range of operating
+scenarios (serial, MPI, job queue submission), we need a way to send a signal to a running OpenMDAO
+model to tell it to return a -2 to pyoptsparse/SNOPT so that we can terminate cleanly.
+
+
+Description
+===========
+
+Sending Signals from the OS
+---------------------------
+Unix provides a selection of signals [1] that are used for interprocess communication. Microsoft
+Windows also provides a smaller set of signals [2] for the same purpose. The discussion here will
+focus on Unix signals, as these cover Unix and OSX systems.  Many of the signals already have
+defined behavior, such as SIGINT which interrupts execution, SIGTERM which terminates the
+process, and SIGSTP which temporarily suspends execution.
+
+Unix provides the `kill` [4] command, which can be used to send any signal to any process.
+Similarly, job queue submission systems also provide a way to forward a specified signal to
+all processes in the job (e.g., Slurm [5] and PBS [6])
+
+It is important to check how an MPI implementation handles signals as well. For example, OpenMPI
+forwards SIGUSR1 and SIGUSR2 to all processes, while HPE MPT does not forward those signals, but
+does forward SIGURG.  Because the signal forwarding behavior is different depending on operating
+system, mpi implementation, and scheduler, the user must retain the ability to choose which
+signal will be used to initiate clean termination.  A reasonable default value is SIGUSR1.
+
+
+Handling Signals in pyOptSparseDriver
+-------------------------------------
+The Python standard library includes a library [3] for dealing with signals, including a way to
+associate a handler function with a particular signal, so that when that signal is detected, the
+handler function is called. When this feature is enabled by the user, a handler will be placed
+on the expected signal. This handler will raise a new exception `UserRequestedException` that
+will be caught in `_objfunc` and `_gradfunc`, where the return code `self.fail` will be set
+to -2.
+
+
+User Interface for Enabling New Capability
+------------------------------------------
+A new option named "user_termination" will be added to the pyOptSparseDriver, with the default
+value of False.  When the user sets this to True, the capability to signal a clean termination will
+be enabled.
+
+A new option named "user_termination_signal" will be added to the pyOptSparseDriver, with the
+default value of `signals.SIGUSR1`.  The user can change this to any other valid signal name or
+number.
+
+
+Changes needed to pyoptsparse
```

Comment:
After discussing this with @johnjasa, we propose the following: instead of adding a new parameter, modify the existing boolean `fail` flag to be an integer, which can encode the exit status. More specifically:

1. Change pyOptSparse's fail flag from a boolean to an integer return code. For now, keep 0/1 so that no behaviour is changed. Some internal MDOlab codes will also need to be updated to reflect this (such as the function wrappers in multipoint).
2. The same change needs to happen in pyOptSparse Driver so that `_objfunc` and `_gradfunc` return integers.
3. Implement signal handling within openMDAO and provide the corresponding exit code to pyOptSparse. I'm not sure how this would be done, especially with MPI but I assume somehow it'll happen within `objfunc` and `gradfunc`.
4. Modify `pySNOPT` to give the correct exit code to SNOPT when evaluations return this exit code.

Does this plan make sense? A few concerns:
1. We need to establish a set of standardized exit codes to communicate between pyOptSparse and openMDAO. This will also need to be clearly documented in pyOptSparse docs.
2. While we transition from boolean to ints, some incompatibility between pyOptSparse and openMDAO may happen. I guess we can add casts to alleviate the problem, until we release new versions.

Reply:
I would propose `fail = -2` as the exit code for user termination, which matches the code that SNOPT receives.

Reply:
I would prefer to stay away from negative exit codes, and come up with a set which is independent of optimizer implementation, since pyOptSparse is a general framework. Within each optimizer wrapper the specific codes can be parsed to the value needed for that optimizer. Something like `fail = 2` would be nice since we will have `0` and `1` for evaluation fail and pass.

Reply:
Sounds good. I will definitely defer to you on that, and implement the return value that general pyoptsparse sees as a user requested termination.

Reply:
Sounds good, thanks. As a more general question, do you think the openMDAO team can handle the implementations on the pyOptSparse side also? I'm afraid we don't currently have the manpower to handle the implementation for some time, since the feature isn't needed by anyone over here.

Reply:
We'll send a PR to the pyoptsparse repo as part of our work.